### PR TITLE
fix(theme): out of stock product will no longer break cart

### PR DESCRIPTION
**Pull request by bartoszherba**
## Description
- fix empty cart issue when one or more products were out of stock
- move cart loading from layout to CartSidebar component for a non-blocking cart data load
- add Out Of Stock flag on a product in minicart
- change apollo client query strategy None -> All so the frontend/composables are now able to process eventual errors.

## Related Issue
#235

## How Has This Been Tested?
With the usage of the test product, toggle its stock status and observe behavior on the frontend.
Sample scenarios:

Test#1
- add a product to the cart
- go to Magento Admin Panel and flag that product as OOS.
- on the VSF, reload page
- observe that label is visible on the product, quantity action input is hidden, the cart was loaded properly

## Screenshots (if appropriate):
Screenshot 2022-01-12 at 14 20 59

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: packages/theme/components/CartSidebar.vue [hunk 1]
```diff
@@ -280,7 +291,7 @@ export default defineComponent({
     const { isAuthenticated } = useUser();
     const { send: sendNotification, notifications } = useUiNotification();
 
-    const products = computed(() => cartGetters.getItems(cart.value));
+    const products = computed(() => cartGetters.getItems(cart.value).filter(Boolean));
```
Comment (Frodigo) [inline packages/theme/components/CartSidebar.vue:294]: You have already filtered items in composable
```suggestion
    const products = computed(() => cartGetters.getItems(cart.value);
```

File: packages/theme/layouts/default.vue [hunk 2]
```diff
@@ -57,16 +56,12 @@ export default defineComponent({
...
     const { load: loadUser } = useUser();
-    const { load: loadCart } = useCart();
 
     const { loadConfiguration } = useMagentoConfiguration();
 
     onSSR(async () => {
       await loadConfiguration();
-      await Promise.all([
-        loadUser(),
-        loadCart(),
```
Comment (Frodigo) [inline packages/theme/layouts/default.vue:68]: So cart is loaded in different place, yeah?

**Review by Frodigo — CHANGES_REQUESTED** [review]

File: packages/theme/layouts/default.vue (hunk 2 above)
Comment (bartoszherba) [reply packages/theme/layouts/default.vue:68]: packages/theme/components/CartSidebar.vue:303

File: packages/theme/components/CartSidebar.vue (hunk 1 above)
Comment (bartoszherba) [reply packages/theme/components/CartSidebar.vue:294]: Sometimes there is null inside, I was not able to track down where is the source.

File: packages/theme/components/__tests__/CartSidebar.spec.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  useCart,
+  useUser,
+} from '@vue-storefront/magento';
+
+import { useUiState } from '~/composables';
+import {
+  render, useCartMock, useUserMock, useUiStateMock, useEmptyCartMock,
+} from '~/test-utils';
+import CartSidebar from '~/components/CartSidebar';
+import stockStatusEnum from '~/enums/stockStatusEnum';
```
Comment (Frodigo) [inline packages/theme/components/__tests__/CartSidebar.spec.js:13]: This is not used, please remove.

File: packages/theme/components/__tests__/CartSidebar.spec.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+useCart.mockReturnValue(useCartMock());
+
+describe('<CartSidebar>', () => {
+  it('should be not visible by default', () => {
+    useUiState.mockReturnValue(useUiStateMock());
+    const { queryByText } = render(CartSidebar);
+    expect(queryByText('My Cart')).toBeNull();
+  });
+
+  describe('If the cart is empty', () => {
```
Comment (Frodigo) [inline packages/theme/components/__tests__/CartSidebar.spec.js:34]: I know that we didn't talk about that, but the common pattern is to use one describe block for one test file. Could you refactor that?  

Besiedes, this is a good read about nesting: https://kentcdodds.com/blog/avoid-nesting-when-youre-testing

File: packages/theme/components/__tests__/CartSidebar.spec.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      it('should render empty state', () => {
+        useUiState.mockReturnValue(useUiStateMock({ isCartSidebarOpen: true }));
+        const { queryByText } = render(CartSidebar);
+        expect(queryByText('Your cart is empty')).toBeTruthy();
+      });
+
+      it('go back button must close the cart sidebar', () => {
+        const uiStateMock = useUiStateMock({ isCartSidebarOpen: true });
+        useUiState.mockReturnValue(uiStateMock);
+        const { queryByText } = render(CartSidebar);
```
Comment (Frodigo) [inline packages/theme/components/__tests__/CartSidebar.spec.js:48]:
```suggestion
        const { getByText } = render(CartSidebar);
```

File: packages/theme/components/__tests__/CartSidebar.spec.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        const closeSidebarBtn = queryByText('Go back shopping');
```
Comment (Frodigo) [inline packages/theme/components/__tests__/CartSidebar.spec.js:49]:
```suggestion
        const closeSidebarBtn = getByText('Go back shopping');
```

File: packages/theme/components/__tests__/CartSidebar.spec.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        expect(closeSidebarBtn).toBeTruthy();
```
Comment (Frodigo) [inline packages/theme/components/__tests__/CartSidebar.spec.js:50]:
```suggestion
        
```

File: packages/theme/components/__tests__/CartSidebar.spec.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+        userEvent.click(closeSidebarBtn);
+        expect(uiStateMock.toggleCartSidebar.mock.calls).toHaveLength(1);
```
Comment (Frodigo) [inline packages/theme/components/__tests__/CartSidebar.spec.js:54]:
```suggestion
         expect(uiStateMock.toggleCartSidebar).toHaveBeenCalledTimes(1);
```

**Review by Frodigo — CHANGES_REQUESTED** [review]

**Comment by Frodigo** [comment]
There are some errors in console: 
Screenshot 2022-01-17 at 07 53 47

File: packages/theme/components/CartSidebar.vue [hunk 9]
```diff
@@ -119,14 +117,25 @@
...
                     <SfQuantitySelector
                       :disabled="loading"
                       :qty="cartGetters.getItemQty(product)"
                       class="sf-collected-product__quantity-selector"
                       @input="updateItemQty({ product, quantity: $event })"
                     />
                   </div>
+                  <SfBadge
+                    v-else
+                    class="color-danger sf-badge__absolute"
```
[the remark this answers is not in the document]
Comment (bartoszherba) [reply packages/theme/components/CartSidebar.vue:133]: I did.

[the remark this answers is not in the document]
Comment (Frodigo) [reply packages/theme/components/CartSidebar.vue:133]: thank you

File: packages/theme/components/__tests__/CartSidebar.spec.js (hunk 4 above)
Comment (bartoszherba) [reply packages/theme/components/__tests__/CartSidebar.spec.js:32]: I do not feel that the test is not readable or hard to understand, or code is hard to follow.
`<CartSidebar>
    ✓ should be not visible by default (17 ms)
    If the cart is empty
      And the cart sidebar is open
        ✓ should render empty state (23 ms)
        ✓ go back button must close the cart sidebar (42 ms)
    If the cart has two products
      ✓ should render two Product Cards (44 ms)
      ✓ should display proper total items value (37 ms)
      ✓ should render "go to checkout" button (35 ms)
      ✓ should render promo code input (27 ms)
      And exactly one product is out of stock
        ✓ should display exactly one out of stock badge (28 ms)`

File: packages/theme/components/__tests__/CartSidebar.spec.js (hunk 5 above)
Comment (bartoszherba) [reply packages/theme/components/__tests__/CartSidebar.spec.js:46]: Query eventually returns null and I want it not the error.

File: packages/theme/components/__tests__/CartSidebar.spec.js (hunk 6 above)
Comment (bartoszherba) [reply packages/theme/components/__tests__/CartSidebar.spec.js:47]: Query eventually returns null and I want it not the error.

File: packages/theme/components/__tests__/CartSidebar.spec.js (hunk 7 above)
Comment (bartoszherba) [reply packages/theme/components/__tests__/CartSidebar.spec.js:48]: I prefer explicit expectation, someone might not know by hearth how another tool is working in depth.

File: packages/theme/components/__tests__/CartSidebar.spec.js (hunk 8 above)
Comment (bartoszherba) [reply packages/theme/components/__tests__/CartSidebar.spec.js:52]: oki

File: packages/theme/components/__tests__/CartSidebar.spec.js (hunk 5 above)
Comment (Frodigo) [reply packages/theme/components/__tests__/CartSidebar.spec.js:46]: I see

File: packages/theme/components/__tests__/CartSidebar.spec.js (hunk 7 above)
Comment (Frodigo) [reply packages/theme/components/__tests__/CartSidebar.spec.js:48]: but `getByText` throws an error when the passed text does not exist in DOM, so the test will break then, the assertion is redundant.

**Review by Frodigo — APPROVED** [review]
